Tests/Realm.Tests/Database/CollectionTests.cs

```diff
@@ -411,6 +414,258 @@ public void Results_GetFiltered_SanityTest()
+            {
+                var matchingObj = _realm.Add(new AllTypesObject { RequiredStringProperty = string.Empty });
+                propInfo.SetValue(matchingObj, castedValues.MatchingVal);
+            });
+
+            var matches = _realm.All<AllTypesObject>().Filter($"{data.PropertyName} = $0", data.NonMatchingValue);
+            Assert.AreEqual(matches.Count(), matchingCount);
+
+            // this method always assumes no more than 1 match, given the structure of StringQueryTestData
+            if (matchingCount == 1)
+            {
+                Assert.AreEqual(propInfo.GetValue(matches.Single()), castedValues.MatchingVal);
+            }
+        }
+
+        [TestCaseSource(nameof(StringQueryTestCases_MismatchingValueTypes_ToThrow))]
+        public void QueryFilter_WithWrongArguments_ShouldThrow(StringQueryTestData data)
+        {
+            var propInfo = typeof(AllTypesObject).GetProperty(data.PropertyName);
+            (object? MatchingVal, object? NonMatchingVal) castedValues = data.ConvertBoxedValueToCorrectType(propInfo.PropertyType);
+            _realm.Write(() =>
+            {
+                var matchingObj = _realm.Add(new AllTypesObject { RequiredStringProperty = string.Empty });
+                propInfo.SetValue(matchingObj, castedValues.MatchingVal);
+            });
+
+            Assert.Throws<RealmException>(() => _realm.All<AllTypesObject>().Filter($"{data.PropertyName} = $0", data.NonMatchingValue), $"Unsupported comparison between type {propInfo.PropertyType.Name} and type {data.NonMatchingValue.GetType().Name}");
+        }
+
+        [Test]
+        public void QueryFilter_WithArgumentsObject_ShouldMatch()
+        {
+            var matchingObj = new PrimaryKeyInt32Object { Int32Property = 42 };
+            var nonMatchingObj = new PrimaryKeyInt32Object { Int32Property = 4 };
+            _realm.Write(() =>
+            {
+                _realm.Add(new ObjectPropertyObject { Object = matchingObj });
+                _realm.Add(new ObjectPropertyObject { Object = nonMatchingObj });
+            });
+
+            var matches = _realm.All<ObjectPropertyObject>().Filter("Object = $0", matchingObj);
+            var foundObj = matches.Single();
+            Assert.AreEqual(matchingObj, foundObj.Object);
+        }
+
+        [Test]
+        public void QueryFilter_WithArgumentsInMemoryObjects_ShouldThrow()
+        {
+            _realm.Write(() =>
+            {
+                _realm.Add(new ObjectPropertyObject { Object = new PrimaryKeyInt32Object { Int32Property = 42 } });
+            });
+
+            Assert.Throws<RealmException>(() => _realm.All<ObjectPropertyObject>().Filter("Object = $0", new PrimaryKeyInt32Object { Int32Property = 42 }), "**put your message here!**");
+        }
+
+        [Test]
+        public void QueryFilter_WithNullArguments_ShouldThrow()
+        {
+            Assert.Throws<ArgumentNullException>(() => _realm.All<Dog>().Filter("Name = $0", null));
```

Comment:
This test passes due to a bug/peculiarity in the way implicit RealmValue conversions happen. Right now `null` is not implicitly converted to `RealmValue` so the compiler will interpret this as the `RealmValue[]` argument being null. When/If we manage to fix the implicit conversion issue, the test will start to fail. To ensure it's testing what you want it to test, it may make sense to assign it to a variable first - `RealmValue[] argumentsArray = null` and then pass that to `Filter`.